### DISCUSSION TITLE
How to update user's password on first login without exposing that we are using Keycloak

### DISCUSSION THREAD

**Discussion by atul-agg** [keycloak/keycloak#34770] [Q&A] [answered]
We are setting temporary password for the user when it is created and changing the password on it's first login. We don't want to expose to the client that we are using keycloak for IAM. I tried using rest calls in POSTMAN but I was not able to update the password for a user. I Identified a few calls that are triggered but unable to get them work in a single flow:

1. **Fetch authenticate user page**
GET http://localhost:8080/realms/testRealm/protocol/openid-connect/auth?response_type=code&client_id=test-client&scope=openid&redirect_uri=http://localhost:8080/realms/testRealm/account/&kc_action=UPDATE_PASSWORD

2. **Provide username and old password inputs**
POST http://localhost:8080/realms/testRealm/login-actions/authenticate?session_code={{session-code}}&execution={{execution-id}}&client_id=test-client&tab_id={{tab-id}}&client_data={{client-data}}
{
    "username": "username",
    "password": "password",
    "credentialId":""
}

3. **Fetch update password page**
GET http://localhost:8080/realms/testRealm/login-actions/required-action?execution=UPDATE_PASSWORD&client_id=test-client&tab_id={{tab-id}}&client_data={{client-data}}

4. **Provide New password and confirm password inputs**
POST http://localhost:8080/realms/testRealm/login-actions/required-action?session_code={{session-code}}&execution=UPDATE_PASSWORD&client_id=test-client&tab_id={{tab-id}}&client_data={{client-data}}
{
   "password-new": "password1",
   "password-confirm": "password1",
   "logout-sessions": "on"
}

Please let me know if anything is missing or incorrect here.

**Comment by atul-agg** [comment]
Any update on this?

**Comment by jbman** [comment, marked as answer]
As you still want to set the new password for a user from the client application, I assume you want to use the password grant.
It doesn't seem to be the right choice to use the Update Password required action, as it is meant to be used for a standard authorization code flow where the user logs in at Keycloak provided forms.

You could set the password without the temporary option. The user provide their intial password which you can recognoze because you prefix it like "change-BiFmdf6c35w". Then you request a password from the user and set it.
